### DISCUSSION TITLE
Cog & WebRTC

### DISCUSSION THREAD

**Discussion by gnulux** [Igalia/cog#480] [General]
Hi, 

I would like to know if WebRTC is supported by by cog browser ? if yes which version i needed to compile/install ? 

B.Regards

**Comment by philn** [comment]
https://wpewebkit.org/about/faq.html#what%E2%80%99s-the-status-regarding-webrtc%3F

**Comment by gnulux** [comment]
Hi ,

Yes , I read this already .
Do you have an example of meta-webkit and Yocto local.conf example in order to build it
with WebRTC support.
I'm trying to make it work on Raspi4 and IMX8  hardware so I tried to
follow all details from there
https://github.com/Igalia/meta-webkit/issues/194 .
But it seems to not work

Regards


Le jeu. 25 août 2022 à 10:59, Philippe Normand ***@***.***> a
écrit :

**Reply by philn** [reply]
Are you trying the libwebrtc approach? What's failing? I can't help if you provide no details :)

About the gstwebrtc backend it's still heavily under development, in 2.38 it won't be much usable, you could do basic p2p calls and use data-channel, but not much more complex like the main WebRTC platforms such as jitsi, teams and so on....

**Reply by gnulux** [reply]
I think libwebrtc is better for me as i have  restriction with version of webkit and Yocto. 

On Raspi4 Yocto  : 
honister build from your wiki with meta-webkit on main branch 
I plan to run a JS application with cog  -P drm (not working) for now,  in wayland (-P fdo working)

Would like to have the same later on IMX8 (Yocto zeus) + meta-webkit (Dunfell branch) as cog does not work on zeus branch (see https://github.com/Igalia/meta-webkit/issues/387).  

For now on raspi4
```
MACHINE = "raspberrypi4-64"
MACHINE_FEATURES:append = " vc4graphics"
GPU_MEM_256 = "128"
GPU_MEM_512 = "196"
GPU_MEM_1024 = "396"
DISTRO_FEATURES:append = " opengl fbdev egl wayland systemd alsa"
EXTRA_IMAGE_FEATURES = "debug-tweaks"
IMAGE_FEATURES:append = " ssh-server-openssh hwcodecs"
PREFERRED_PROVIDER_virtual/wpebackend = "wpebackend-fdo"
IMAGE_INSTALL:append = " wpewebkit cog nodejs net-tools htop chrony"
VIDEO_CAMERA = "1"

PREFERRED_VERSION_wpebackend-fdo ?= "1.12.0"
PREFERRED_VERSION_wpewebkit ?= "2.34.7"
PREFERRED_VERSION_libwpe ?= "1.12.0"
PREFERRED_VERSION_cog ?= "0.10.1"
PREFERED_VERSION_nodejs ?= "16.14.2"

PACKAGECONFIG:append:pn-wpewebkit = " deviceorientation encryptedmedia fetchapi gamepad gst_gl indexeddb mediasource video webaudio webcrypto woff2 2dcanvas remote-inspector webrtc"
PACKAGECONFIG:append:pn-gstreamer1.0-plugins-ugly = " mpeg2dec x264"
PACKAGECONFIG:append:pn-gstreamer1.0-plugins-good = " cairo jpeg libpng libv4l2 soup v4l2"
PACKAGECONFIG:append:pn-gstreamer1.0-plugins-bad = " webrtc egl gles2 opengl openjpeg"
LICENSE_FLAGS_ACCEPTED += "commercial"

BB_NUMBER_THREADS = "6"
```
It builds without any issue, but when i launch ``` cog -C config.txt -P fdo https://html5test.com ``` all elements in Peer to Peer section are still disabled. The other features like "deviceorientation" are enabled as required. 
with ```cog -C config.txt -P fdo https://people.igalia.com/clopez/wkbug/webrtc/mediastream-detect.html```
MediaStream detection broken (unknown issue... see JS console)

on console : 
```
CONSOLE JS ERROR TypeError: navigator.mediaDevices is not an Object. (evaluating  ''getUserMedia' in navigator.MediaDevices') 
```

config.txt
```
[websettings]
enable-write-console-messages-to-stdout = true
enable-media-stream = true
set-permissions = all
```

**Reply by philn** [reply]
As mentioned in the ticket linked previously you also need:

1. Build WPE from a git checkout
2. `experimental-features` enabled
